helix-core/src/main/java/org/apache/helix/messaging/handling/HelixTaskExecutor.java

```diff
@@ -827,7 +833,11 @@ public void onMessage(String instanceName, List<Message> messages,
     // and terminate all the thread pools
     // TODO: see if we should have a separate notification call for resetting
     if (changeContext.getType() == Type.FINALIZE) {
-      reset();
+      if (_isShuttingDown) {
```

Comment:
where is the flag being set?

Reply:
https://github.com/apache/helix/blob/132715785e90803ad8991da491f4621db1668fb8/helix-core/src/main/java/org/apache/helix/messaging/handling/HelixTaskExecutor.java#L1454